Cognite.Simulator.Tests/UtilsTests/ConnectorRuntimeUnitTest.cs

```diff
@@ -0,0 +1,132 @@
+            { uri => uri.Contains("/extpipes"), (MockExtPipesEndpoint, 0, null) },
+            { uri => uri.EndsWith("/simulators/list") || uri.EndsWith("/simulators"), (MockSimulatorsEndpoint, 0, null) },
+            { uri => uri.Contains("/simulators/integrations"), (MockSimulatorsIntegrationsEndpoint, 0, null) },
+            { uri => uri.Contains("/simulators/routines/revisions/list"), (MockSimulatorRoutineRevEndpoint, 0, 1) }
+        };
+
+        // We need to mock the HttpClientFactory to return the mocked responses
+        // First few requests return the mocked responses, then we return a 403 Forbidden
+        // This should cause a "soft" restart of the connector
+        [Fact]
+        public async Task TestConnectorRuntimeWithRestart()
+        {
+            WriteConfig();
+
+            var logger = LoggingUtils.GetDefault();
+            var mockedLogger = new Mock<ILogger<DefaultConnectorRuntime<DefaultAutomationConfig, DefaultModelFilestate, DefaultModelFileStatePoco>>>();
+
+            using var cts = new CancellationTokenSource();
+            cts.CancelAfter(TimeSpan.FromSeconds(3));
+
+            var mocks = GetMockedHttpClientFactory(mockRequestsAsync(endpointMappings));
+            var mockHttpMessageHandler = mocks.handler;
+            var mockFactory = mocks.factory;
+
+            DefaultConnectorRuntime<DefaultAutomationConfig, DefaultModelFilestate, DefaultModelFileStatePoco>.ConfigureServices = (services) => {
+                services.AddScoped<ISimulatorClient<DefaultModelFilestate, SimulatorRoutineRevision>, EmptySimulatorAutomationClient>();
+                services.AddSingleton(mockFactory.Object);
+                services.AddSingleton(mockedLogger.Object);
+            };
+            DefaultConnectorRuntime<DefaultAutomationConfig, DefaultModelFilestate, DefaultModelFileStatePoco>.ConnectorName = "Empty";
+            DefaultConnectorRuntime<DefaultAutomationConfig, DefaultModelFilestate, DefaultModelFileStatePoco>.SimulatorDefinition = SeedData.SimulatorCreate;
+            
+            try {
+                await DefaultConnectorRuntime<DefaultAutomationConfig, DefaultModelFilestate, DefaultModelFileStatePoco>.Run(logger, cts.Token).ConfigureAwait(false);
+            } catch (OperationCanceledException) {}
+        
+            // Check the logs, it should first succeed on the startup, then fail and restart
+            VerifyLog(mockedLogger, LogLevel.Information, "Starting the connector...", Times.Once());
+            VerifyLog(mockedLogger, LogLevel.Information, "Connector can reach CDF!", Times.Once());
+            VerifyLog(mockedLogger, LogLevel.Debug, "Updating simulator definition", Times.Once());
+            VerifyLog(mockedLogger, LogLevel.Error, "Request to CDF failed with code 403", Times.Once(), true);
+            VerifyLog(mockedLogger, LogLevel.Warning, "Restarting connector in 5 seconds", Times.AtLeastOnce());
```

Comment:
Not sharing any review feedback just leaving a comment to appreciate how good this solution looks.

Reply:
`VerifyLog` specifically is something @abdullah-cognite made, so I am not taking a credit for that!
I did the HTTP mocking part 😇